File: java/src/test/java/com/turbotoken/EncodingTest.java
package com.turbotoken;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for turbotoken Java bindings.
 * Tests marked with @EnabledIfEnvironmentVariable require the native library.
 */
class EncodingTest {

    @Test
    void registryListEncodingNames() {
        List<String> names = Registry.listEncodingNames();
        assertNotNull(names);
        assertTrue(names.contains("cl100k_base"));
        assertTrue(names.contains("o200k_base"));
        assertTrue(names.contains("p50k_base"));
        assertTrue(names.contains("r50k_base"));
        assertTrue(names.contains("gpt2"));
        assertTrue(names.contains("p50k_edit"));
        assertTrue(names.contains("o200k_harmony"));
        assertEquals(7, names.size());
    }

    @Test
    void registryGetEncodingSpec() {
        Registry.EncodingSpec spec = Registry.getEncodingSpec("cl100k_base");
        assertEquals("cl100k_base", spec.getName());
        assertEquals(100277, spec.getNVocab());
        assertEquals(100257, spec.getEotToken());
        assertTrue(spec.getRankFileUrl().contains("cl100k_base.tiktoken"));
    }

    @Test
    void registryGetEncodingSpecUnknown() {
        assertThrows(IllegalArgumentException.class, () -> Registry.getEncodingSpec("nonexistent"));
    }

    @Test
    void registryModelToEncoding() {
        assertEquals("o200k_base", Registry.modelToEncoding("gpt-4o"));
        assertEquals("cl100k_base", Registry.modelToEncoding("gpt-4"));
        assertEquals("r50k_base", Registry.modelToEncoding("davinci"));
        assertEquals("p50k_base", Registry.modelToEncoding("text-davinci-003"));
    }

    @Test
    void registryModelToEncodingPrefix() {
        assertEquals("o200k_base", Registry.modelToEncoding("gpt-4o-2024-08-06"));
        assertEquals("cl100k_base", Registry.modelToEncoding("gpt-4-turbo-preview"));
        assertEquals("o200k_base", Registry.modelToEncoding("o1-preview"));
    }

    @Test
    void registryModelToEncodingUnknown() {
        assertThrows(IllegalArgumentException.class, () -> Registry.modelToEncoding("unknown-model-xyz"));
    }

    @Test
    void turboTokenListEncodings() {
        List<String> names = TurboToken.listEncodingNames();
        assertNotNull(names);
        assertEquals(7, names.size());
    }

    @Test
    void chatTemplateResolve() {
        ChatTemplate.Template template = ChatTemplate.resolve(ChatTemplate.TemplateMode.TURBOTOKEN_V1);
        assertNotNull(template.getMessagePrefix());
        assertNotNull(template.getMessageSuffix());
        assertNotNull(template.getAssistantPrefix());
    }

    @Test
    void chatTemplateFormat() {
        List<ChatTemplate.ChatMessage> messages = List.of(
            new ChatTemplate.ChatMessage("system", "You are helpful."),
            new ChatTemplate.ChatMessage("user", "Hello!")
        );
        ChatTemplate.ChatOptions options = new ChatTemplate.ChatOptions();
        String formatted = ChatTemplate.formatMessages(messages, options);
        assertTrue(formatted.contains("system"));
        assertTrue(formatted.contains("You are helpful."));
        assertTrue(formatted.contains("user"));
        assertTrue(formatted.contains("Hello!"));
    }

    /* ── Native tests (require library loaded) ───────────────────────── */

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    void encodeDecodeRoundTrip() {
        Encoding enc = TurboToken.getEncoding("cl100k_base");
        String original = "hello world";
        int[] tokens = enc.encode(original);
        assertNotNull(tokens);
        assertTrue(tokens.length > 0);
        String decoded = enc.decode(tokens);
        assertEquals(original, decoded);
    }

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    void count() {
        Encoding enc = TurboToken.getEncoding("cl100k_base");
        int count = enc.count("hello world");
        assertTrue(count > 0);
        int[] tokens = enc.encode("hello world");
        assertEquals(tokens.length, count);
    }

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    void getEncodingByModel() {
        Encoding enc = TurboToken.getEncodingForModel("gpt-4o");
        assertEquals("o200k_base", enc.getName());
        int[] tokens = enc.encode("test");
        assertNotNull(tokens);
        assertTrue(tokens.length > 0);
    }
}

File: wrappers/kotlin/src/test/kotlin/com/turbotoken/EncodingTest.kt
package com.turbotoken

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Tests for turbotoken Kotlin bindings.
 */
class KEncodingTest {

    @Test
    fun `registry lists all encoding names`() {
        val names = EncodingRegistry.listEncodingNames()
        assertNotNull(names)
        assertTrue(names.contains("cl100k_base"))
        assertTrue(names.contains("o200k_base"))
        assertTrue(names.contains("p50k_base"))
        assertTrue(names.contains("r50k_base"))
        assertTrue(names.contains("gpt2"))
        assertTrue(names.contains("p50k_edit"))
        assertTrue(names.contains("o200k_harmony"))
        assertEquals(7, names.size)
    }

    @Test
    fun `registry get encoding spec`() {
        val spec = EncodingRegistry.getEncodingSpec("cl100k_base")
        assertEquals("cl100k_base", spec.name)
        assertEquals(100277, spec.nVocab)
        assertEquals(100257, spec.eotToken)
    }

    @Test
    fun `registry get encoding spec or null returns null for unknown`() {
        val spec = EncodingRegistry.getEncodingSpecOrNull("nonexistent")
        assertNull(spec)
    }

    @Test
    fun `registry model to encoding`() {
        assertEquals("o200k_base", EncodingRegistry.modelToEncoding("gpt-4o"))
        assertEquals("cl100k_base", EncodingRegistry.modelToEncoding("gpt-4"))
        assertEquals("r50k_base", EncodingRegistry.modelToEncoding("davinci"))
    }

    @Test
    fun `registry model to encoding or null returns null for unknown`() {
        val enc = EncodingRegistry.modelToEncodingOrNull("unknown-model-xyz")
        assertNull(enc)
    }

    @Test
    fun `registry model to encoding prefix match`() {
        assertEquals("o200k_base", EncodingRegistry.modelToEncoding("gpt-4o-2024-08-06"))
        assertEquals("cl100k_base", EncodingRegistry.modelToEncoding("gpt-4-turbo-preview"))
    }

    @Test
    fun `turbotoken list encoding names`() {
        val names = TurboTokenK.listEncodingNames()
        assertNotNull(names)
        assertEquals(7, names.size)
    }

    @Test
    fun `chat template resolve`() {
        val template = resolveTemplate(KTemplateMode.TurbotokenV1)
        assertNotNull(template.messagePrefix)
        assertNotNull(template.messageSuffix)
        assertNotNull(template.assistantPrefix)
    }

    @Test
    fun `chat template format messages`() {
        val messages = listOf(
            KChatMessage(role = "system", content = "You are helpful."),
            KChatMessage(role = "user", content = "Hello!")
        )
        val formatted = formatMessages(messages)
        assertTrue(formatted.contains("system"))
        assertTrue(formatted.contains("You are helpful."))
        assertTrue(formatted.contains("user"))
        assertTrue(formatted.contains("Hello!"))
    }

    @Test
    fun `chat message with name`() {
        val msg = KChatMessage(role = "user", content = "Hi", name = "Alice")
        assertEquals("user", msg.role)
        assertEquals("Hi", msg.content)
        assertEquals("Alice", msg.name)
    }

    /* ── Native tests (require library loaded) ───────────────────────── */

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    fun `encode decode round trip`() {
        val enc = TurboTokenK.getEncoding("cl100k_base")
        val original = "hello world"
        val tokens = enc.encode(original)
        assertNotNull(tokens)
        assertTrue(tokens.isNotEmpty())
        val decoded = enc.decode(tokens)
        assertEquals(original, decoded)
    }

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    fun `count tokens`() {
        val enc = TurboTokenK.getEncoding("cl100k_base")
        val count = enc.count("hello world")
        assertTrue(count > 0)
        val tokens = enc.encode("hello world")
        assertEquals(tokens.size, count)
    }

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    fun `extension function token count`() {
        val enc = TurboTokenK.getEncoding("cl100k_base")
        val count = "hello world".tokenCount(enc)
        assertTrue(count > 0)
    }

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    fun `extension function encode`() {
        val enc = TurboTokenK.getEncoding("cl100k_base")
        val tokens = "hello world".encode(enc)
        assertNotNull(tokens)
        assertTrue(tokens.isNotEmpty())
    }

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    fun `result variant captures errors gracefully`() {
        val enc = TurboTokenK.getEncoding("cl100k_base")
        val result = enc.encodeResult("test string")
        assertTrue(result.isSuccess)
        assertNotNull(result.getOrNull())
    }

    @Test
    @EnabledIfEnvironmentVariable(named = "TURBOTOKEN_NATIVE_LIB", matches = ".+")
    fun `get encoding for model`() {
        val enc = TurboTokenK.getEncodingForModel("gpt-4o")
        assertEquals("o200k_base", enc.name)
        val tokens = enc.encode("test")
        assertNotNull(tokens)
        assertTrue(tokens.isNotEmpty())
    }
}
